syntax = "proto3";

package arduino_trader.universe;

import "common/common.proto";
import "common/security.proto";

// Universe Service - Handles security universe and market data
service UniverseService {
  // Get all securities in universe
  rpc GetUniverse(GetUniverseRequest) returns (GetUniverseResponse);

  // Get a specific security
  rpc GetSecurity(GetSecurityRequest) returns (GetSecurityResponse);

  // Search securities
  rpc SearchSecurities(SearchSecuritiesRequest) returns (SearchSecuritiesResponse);

  // Sync prices from external APIs
  rpc SyncPrices(SyncPricesRequest) returns (stream SyncPricesUpdate);

  // Sync fundamentals
  rpc SyncFundamentals(SyncFundamentalsRequest) returns (stream SyncFundamentalsUpdate);

  // Get market data for a security
  rpc GetMarketData(GetMarketDataRequest) returns (GetMarketDataResponse);

  // Add security to universe
  rpc AddSecurity(AddSecurityRequest) returns (AddSecurityResponse);

  // Remove security from universe
  rpc RemoveSecurity(RemoveSecurityRequest) returns (RemoveSecurityResponse);

  // Health check
  rpc HealthCheck(common.Empty) returns (common.HealthCheckResponse);
}

// Get universe request
message GetUniverseRequest {
  bool tradable_only = 1;
  string sector = 2;  // Optional filter
  string exchange = 3;  // Optional filter
  int32 limit = 4;
  int32 offset = 5;
}

// Get universe response
message GetUniverseResponse {
  repeated common.Security securities = 1;
  int32 total = 2;
}

// Get security request
message GetSecurityRequest {
  string isin = 1;
  string symbol = 2;  // Alternative to ISIN
}

// Get security response
message GetSecurityResponse {
  bool found = 1;
  common.Security security = 2;
}

// Search securities request
message SearchSecuritiesRequest {
  string query = 1;
  int32 limit = 2;
}

// Search securities response
message SearchSecuritiesResponse {
  repeated common.Security securities = 1;
  int32 total_matches = 2;
}

// Sync prices request
message SyncPricesRequest {
  repeated string isins = 1;  // Empty = sync all
  bool force_refresh = 2;
}

// Sync prices update (streaming)
message SyncPricesUpdate {
  int32 progress_pct = 1;
  int32 synced = 2;
  int32 failed = 3;
  int32 total = 4;
  string current_isin = 5;
  bool complete = 6;
  string error = 7;
}

// Sync fundamentals request
message SyncFundamentalsRequest {
  repeated string isins = 1;  // Empty = sync all
  bool force_refresh = 2;
}

// Sync fundamentals update (streaming)
message SyncFundamentalsUpdate {
  int32 progress_pct = 1;
  int32 synced = 2;
  int32 failed = 3;
  int32 total = 4;
  string current_isin = 5;
  bool complete = 6;
  string error = 7;
}

// Get market data request
message GetMarketDataRequest {
  string isin = 1;
  int32 days = 2;  // Historical data
}

// Get market data response
message GetMarketDataResponse {
  string isin = 1;
  common.Security security = 2;
  repeated PriceBar history = 3;
  MarketMetrics metrics = 4;
}

// Price bar (OHLCV)
message PriceBar {
  common.Timestamp date = 1;
  double open = 2;
  double high = 3;
  double low = 4;
  double close = 5;
  int64 volume = 6;
  double adjusted_close = 7;
}

// Market metrics
message MarketMetrics {
  double beta = 1;
  double volatility = 2;
  double average_volume = 3;
  double market_cap = 4;
  double pe_ratio = 5;
  double dividend_yield = 6;
  double day_change_pct = 7;
  common.Money day_change = 8;
  common.Money week_52_high = 9;
  common.Money week_52_low = 10;
}

// Add security request
message AddSecurityRequest {
  string isin = 1;
  string symbol = 2;
  string name = 3;
  string exchange = 4;
  bool sync_immediately = 5;
}

// Add security response
message AddSecurityResponse {
  bool success = 1;
  common.Security security = 2;
  string message = 3;
}

// Remove security request
message RemoveSecurityRequest {
  string isin = 1;
  string reason = 2;
}

// Remove security response
message RemoveSecurityResponse {
  bool success = 1;
  string message = 2;
}
